src/client/connstring.rs

```diff
@@ -0,0 +1,311 @@
+use std::ascii::AsciiExt;
+use std::collections::HashMap;
+use std::vec::Vec;
+
+pub const DEFAULT_PORT: u16 = 27017;
+pub const URI_SCHEME: &'static str = "mongodb://";
+
+/// Encapsulates the hostname and port of a host.
+pub struct Host {
+    pub host_name: String,
+    pub ipc: String,
+    pub port: u16,
+}
+
+impl Host {
+    // Creates a new Host struct.
+    fn new(host_name: String, port: u16) -> Host {
+        Host {
+            host_name: host_name,
+            port: port,
+            ipc: String::new(),
+        }
+    }
+
+    fn new_ipc(ipc: String) -> Host {
+        Host {
+            host_name: String::new(),
+            port: DEFAULT_PORT,
+            ipc: ipc,
+        }
+    }
+
+    pub fn is_ipc(&self) -> bool {
+        self.ipc.len() > 0
+    }
+}
+
+/// Encapsulates the options and read preference tags of a MongoDB connection.
+pub struct ConnectionOptions {
+    pub options: HashMap<String, String>,
+    pub read_pref_tags: Vec<String>,
+}
+
+impl ConnectionOptions {
+    /// Creates a new ConnectionOptions struct.
+    pub fn new(options: HashMap<String, String>, read_pref_tags: Vec<String>) -> ConnectionOptions {
+        ConnectionOptions {
+            options: options,
+            read_pref_tags: read_pref_tags,
+        }
+    }
+
+    /// Retrieves an option using a borrowed String
+    pub fn get(&self, key: &String) -> Option<&String> {
```

Comment:
`&String` is almost never what you want. It's a pointer to a pointer, and you don't need the second indirection. Both parts of this signature should probably be `&str`.

`HashMap` specifically supports looking up `String` keys with `&str`: http://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.get

    fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
        where K: Borrow<Q>,
                  Q: Hash + Eq

This type signature is _intense_, but the important part is `K: Borrow<Q>`, in other words, the key's type supports `Borrow` for the argument's type. And `impl Borrow<str> for String` is in the standard library. So it works: http://is.gd/ywiEUu